server/http4s-server/src/main/scala/sttp/tapir/server/http4s/Http4sServerToHttpInterpreter.scala

```diff
@@ -0,0 +1,139 @@
+package sttp.tapir.server.http4s
+
+import cats.data.{Kleisli, OptionT}
+import cats.effect.{Concurrent, ContextShift, Sync, Timer}
+import cats.implicits._
+import cats.~>
+import fs2.Pipe
+import fs2.concurrent.Queue
+import org.http4s._
+import org.http4s.server.websocket.WebSocketBuilder
+import org.http4s.websocket.WebSocketFrame
+import org.log4s.{Logger, getLogger}
+import org.typelevel.ci.CIString
+import sttp.capabilities.WebSockets
+import sttp.capabilities.fs2.Fs2Streams
+import sttp.tapir.Endpoint
+import sttp.tapir.integ.cats.CatsMonadError
+import sttp.tapir.model.ServerResponse
+import sttp.tapir.server.ServerEndpoint
+import sttp.tapir.server.interpreter.{BodyListener, ServerInterpreter}
+
+import scala.reflect.ClassTag
+
+trait Http4sServerToHttpInterpreter[F[_], G[_]] {
+
+  implicit def gs: Sync[G]
+  implicit def gcs: ContextShift[G]
+  implicit def fs: Concurrent[F]
+  implicit def fcs: ContextShift[F]
+  implicit def timer: Timer[F]
+
+  def fToG: F ~> G
+  def gToF: G ~> F
+
+  def http4sServerOptions: Http4sServerOptions[F, G] = Http4sServerOptions.default[F, G]
+
+  def toHttp[I, E, O](
+      e: Endpoint[I, E, O, Fs2Streams[F] with WebSockets]
+  )(logic: I => G[Either[E, O]]): Http[OptionT[G, *], F] = toHttp(e.serverLogic(logic))
+
+  def toHttpRecoverErrors[I, E, O](
+      e: Endpoint[I, E, O, Fs2Streams[F] with WebSockets]
+  )(logic: I => G[O])(implicit
+      eIsThrowable: E <:< Throwable,
+      eClassTag: ClassTag[E]
+  ): Http[OptionT[G, *], F] = toHttp(e.serverLogicRecoverErrors(logic))
+
+  //
+
+  def toHttp[I, E, O](se: ServerEndpoint[I, E, O, Fs2Streams[F] with WebSockets, G]): Http[OptionT[G, *], F] = toHttp(List(se))(fToG)(gToF)
+
+  def toHttp(
+      serverEndpoints: List[ServerEndpoint[_, _, _, Fs2Streams[F] with WebSockets, G]]
+  )(fToG: F ~> G)(gToF: G ~> F): Http[OptionT[G, *], F] = {
```

Comment:
Hi @adamw! Can you explain purpose of `gToF: G ~> F`? Just `toHttp: Http[OptionT[G, *], F]` was added by https://github.com/softwaremill/tapir/pull/683 and in original it has only `fToG: F ~> G` but addition of  `gToF: G ~> F` absolutely destroy the original idea.

Reply:
in `Http4sBodyListener`, we need to know when the body has finished streaming - hence we need a way to translate the effect of the callback (in which the response is returned) to the effect in which the body is returned.

If you'd have a better idea how to model this, I'm open to suggestions :)

Reply:
There are a lot of changes was made. I'll try to see what you mean ASAP :)

Reply:
I've seen than the `ServerInterpriter` was generalized and as as result `ServerInterpriter` should support transformation `F ~> G` but unfortunately it didn't happen. It was an mistake to  propagate `G` in the `Http4sBodyListener` and others  because `F` is an effect-full computation not `G`. `G` is just a monad needed only in the `Http[OptionT[G, *], F]`. It would be better to remove `G` and `F ~> G` at all because it only confuse :)

Reply:
@leoniv yeah you're right, unfortunately keeping tapir general & abstract enough to support multiple frameworks has its price

see https://github.com/softwaremill/tapir/pull/2042